core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -818,6 +858,45 @@ private static AnnotatedValueResolver ofInjectableTypes0(String name, AnnotatedE
         };
     }
 
+    /**
+     * Returns an attribute resolver which retrieves a value specified by {@code attrKeys}
+     * from the {@link RequestContext}.
+     */
+    private static BiFunction<AnnotatedValueResolver, ResolverContext, Object>
+    attributeResolver(Iterable<AttributeKey<?>> attrKeys) {
+        return (resolver, ctx) -> {
+            Object lastInvalidValue = null;
+            String lastAttrKeyName = null;
+            Class<?> targetType = resolver.rawType();
+            if (targetType.isPrimitive()) {
+                targetType = Primitives.wrap(targetType);
+            }
+
+            for (AttributeKey<?> attrKey : attrKeys) {
+                final Object value = ctx.context.attr(attrKey);
+                if (value != null) {
+                    final boolean isValidType = targetType.isInstance(value);
+                    if (isValidType) {
+                        return value;
+                    } else {
+                        lastInvalidValue = value;
+                        lastAttrKeyName = attrKey.name();
+                    }
```

Comment:
Maybe we should throw immediately rather than trying the next candidate to avoid ambiguity?

Reply:
@trustin nim, thanks for your comment 🙇‍♂️ 
IMHO, we should not throw immediately...! 

We have two attribute keys when `prefix()` is not specified. 
If we throw immediatley, i think we should remove `builder.add(AttributeKey.valueOf(name));`. 
In other words, if we want to hold `builder.add(AttributeKey.valueOf(name))`, we don't  throw immediately.

```java
// AnnotatedValueResolver.java
...
    private static AnnotatedValueResolver ofAttribute(...){
...

        if (attr.prefix() != Attribute.class) {
...
        } else {
            final Class<?> serviceClass = ((Parameter) annotatedElement).getDeclaringExecutable()
                                                                        .getDeclaringClass();
            builder.add(AttributeKey.valueOf(serviceClass, name));
            builder.add(AttributeKey.valueOf(name));
        }
```
https://github.com/chickenchickenlove/armeria/blob/0dd3e7767c880f99b13c19d8599bc0cceeab19cd/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java#L644-L651

Reply:
I think we should throw because otherwise it will fail to tell the user about the type casting errors and silently fall back to the second candidate which might not be an expected behavior. Ignoring type casting issue silently is in general a surprising behavior, so I'd avoid doing that.

Reply:
I agree. 
Also, if we throw immediately, code fragments below is still meaningful.
```java
            builder.add(AttributeKey.valueOf(serviceClass, name));
            builder.add(AttributeKey.valueOf(name));
```
Because, just fail to find this value  `builder.add(AttributeKey.valueOf(serviceClass, name));`, not a fail to cast. 

Thus, i fixed it like this.
1. throw error immediately if fail to cast
2. remain codes below as it is.
```java
            builder.add(AttributeKey.valueOf(serviceClass, name));
            builder.add(AttributeKey.valueOf(name));
```

Reply:
Thanks! Looks good now.